tests/test_nanoevents_vector.py

```diff
@@ -532,9 +538,9 @@ def test_inherited_method_transpose(lcoord, threecoord, twocoord):
     assert record_arrays_equal(a.delta_phi(c), c.delta_phi(a))
     assert record_arrays_equal(b.delta_phi(c), c.delta_phi(b))
 
-    assert record_arrays_equal(a - b, -(b - a))
-    assert record_arrays_equal(a - c, -(c - a))
-    assert record_arrays_equal(b - c, -(c - b))
+    assert record_arrays_equal((a - b).to_Vector3D(), -(b - a))
+    assert record_arrays_equal((a - c).to_Vector2D(), -(c - a))
+    assert record_arrays_equal((b - c).to_Vector2D(), -(c - b))
```

Comment:
`a` is a 4-vector, `b` a 3-vector, and `c` a 2-vector. So I guess the type demotion behavior depends on only the first argument of the infix operator here? Is this expected?

Reply:
@saransh-cpp do we expect this (3D to 2D vector demotion / promotion)?

Reply:
Yes, the behavior is uniform across all backends.

Reply:
In that you expect when you do `==` between two vectors you have to cast them to the same dimensionality first?


@nsmith- what was the expected behavior in nanoevents.methods.vector?

Reply:
Yes, for `==` the vectors should have same dimensions. `to_Vector*D` or `to_{coordinate_names}` methods can be used to change the dimensions. `to_Vector*D` additionally allows users to update the newly created coordinate (when the vector's dimension is promoted).

Reply:
My opinion is you should promote to the highest vector in the operation and then cast down in dimensionality as you need. That has the most reasonable behavior from my perspective.

However right now we do neither!

Reply:
I disagree. By promoting the 2-vector to a 3- or 4-vector, you're assuming a value of 0 for the missing dimension(s). If someone wants to add MET to a 4-vector they should have to explicitly upcast it from a 2-vector and define explicitly the pseudorapidity and mass.
I also don't think that `a - c` and `c - a` should result in different types.

Reply:
`a - c` vs. `c - a` implies some sort of promotion or demotion has to happen. So you are asking that we always demote?

Then common patterns like `some_4_vector + 2_vector_spatial_rotation` no longer result in a 4-vector, which would be very confusing.

Reply:
Yes I think we should always demote. I don't understand what you mean by adding a rotation. A rotation depends on the current vector. Anyway the current coffea vectors demote always:
```
>>> a-c
<TwoVectorArray [{x: 20, y: -22}, ..., {x: 15, ...}] type='3 * TwoVector[x:...'>
>>> c-a
<TwoVectorArray [{x: -20, y: 22}, ..., {x: -15, ...}] type='3 * TwoVector[x...'>
>>> a+c
<TwoVectorArray [{x: 0, y: 2}, {...}, {x: 45, ...}] type='3 * TwoVector[x: ...'>
>>> c+a
<TwoVectorArray [{x: 0, y: 2}, {...}, {x: 45, ...}] type='3 * TwoVector[x: ...'>
>>> b-c
<TwoVectorArray [{x: 0, y: -22}, ..., {x: -45, ...}] type='3 * TwoVector[x:...'>
>>> b+c
<TwoVectorArray [{x: -20, y: 2}, ..., {x: -15, ...}] type='3 * TwoVector[x:...'>
>>> c-b
<TwoVectorArray [{x: 0, y: 22}, ..., {x: 45, ...}] type='3 * TwoVector[x: f...'>
>>> c+b
<TwoVectorArray [{x: -20, y: 2}, ..., {x: -15, ...}] type='3 * TwoVector[x:...'>
>>> a+b
<ThreeVectorArray [{x: 0, y: -20, z: 10}, ..., {...}] type='3 * ThreeVector...'>
```

Reply:
Arg, morning brain glitch(re: rotations).

Reply:
@Saransh-cpp I think this is just a bug in vector: https://github.com/scikit-hep/vector/issues/412